chainservice/builder.go

```diff
@@ -0,0 +1,689 @@
+		factory.RegistryOption(builder.cs.registry),
+		factory.DefaultTriePatchOption(),
+	)
+}
+
+func (builder *Builder) buildElectionCommittee() error {
+	ec, err := builder.createElectionCommittee()
+	if err != nil {
+		return errors.Wrapf(err, "failed to create election committee")
+	}
+	if ec != nil {
+		builder.cs.electionCommittee = ec
+		builder.cs.lifecycle.Add(ec)
+	}
+	return nil
+}
+
+func (builder *Builder) createElectionCommittee() (committee.Committee, error) {
+	if builder.cs.electionCommittee != nil {
+		return builder.cs.electionCommittee, nil
+	}
+	if !builder.cfg.Genesis.EnableGravityChainVoting {
+		return nil, nil
+	}
+	cfg := builder.cfg
+
+	committeeConfig := builder.cfg.Chain.Committee
+	if committeeConfig.GravityChainStartHeight == 0 {
+		return nil, nil
+	}
+	committeeConfig.GravityChainStartHeight = builder.cfg.Genesis.GravityChainStartHeight
+	committeeConfig.GravityChainCeilingHeight = cfg.Genesis.GravityChainCeilingHeight
+	committeeConfig.GravityChainHeightInterval = cfg.Genesis.GravityChainHeightInterval
+	committeeConfig.RegisterContractAddress = cfg.Genesis.RegisterContractAddress
+	committeeConfig.StakingContractAddress = cfg.Genesis.StakingContractAddress
+	committeeConfig.VoteThreshold = cfg.Genesis.VoteThreshold
+	committeeConfig.ScoreThreshold = "0"
+	committeeConfig.StakingContractAddress = cfg.Genesis.StakingContractAddress
+	committeeConfig.SelfStakingThreshold = cfg.Genesis.SelfStakingThreshold
+
+	arch, err := committee.NewArchive(
+		cfg.Chain.GravityChainDB.DbPath,
+		cfg.Chain.GravityChainDB.NumRetries,
+		committeeConfig.GravityChainStartHeight,
+		committeeConfig.GravityChainHeightInterval,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return committee.NewCommittee(arch, committeeConfig)
+}
+
+func (builder *Builder) buildActionPool() error {
+	if builder.cs.actpool == nil {
+		ac, err := actpool.NewActPool(builder.cs.factory, builder.cfg.ActPool)
+		if err != nil {
+			return errors.Wrap(err, "failed to create actpool")
+		}
+		builder.cs.actpool = ac
+	}
+	// Add action validators
+	builder.cs.actpool.AddActionEnvelopeValidators(
+		protocol.NewGenericValidator(builder.cs.factory, accountutil.AccountState),
+	)
+
+	return nil
+}
+
+func (builder *Builder) buildBlockDAO(forTest bool) error {
+	if builder.cs.blockdao != nil {
+		return nil
+	}
+
+	var indexers []blockdao.BlockIndexer
+	indexers = append(indexers, builder.cs.factory)
+	if !builder.cfg.Chain.EnableAsyncIndexWrite {
+		if builder.cs.bfIndexer != nil {
+			indexers = append(indexers, builder.cs.bfIndexer)
+		}
+		if builder.cs.indexer != nil {
+			indexers = append(indexers, builder.cs.indexer)
+		}
+	}
+	if forTest {
+		builder.cs.blockdao = blockdao.NewBlockDAOInMemForTest(indexers)
+	} else {
+		dbConfig := builder.cfg.DB
+		dbConfig.DbPath = builder.cfg.Chain.ChainDBPath
+		dbConfig.CompressLegacy = builder.cfg.Chain.CompressBlock
+
+		builder.cs.blockdao = blockdao.NewBlockDAO(indexers, dbConfig)
+	}
+
```

Comment:
would you mind remove line before `return`? It's 2 diff styles, IMO no new line looks OK, would like to keep it the same

Reply:
I won't to leave an empty line before return.